src/emulator/cpu.ts

```diff
@@ -336,119 +336,130 @@ export class Cpu {
         }
     }
 
-    private getArg1(instruction: Instruction) {
-        switch (instruction.addressingMode) {
+    private getArg(par: OperationParameter): number {
+        switch (par.addressingMode) {
             case AddressingMode.imm8:
-            case AddressingMode.imm8_reg8:
                 return this.bus.read((this.state.p + 1) & 0xffff);
 
-            case AddressingMode.imm16:
-                return this.bus.read16((this.state.p + 1) & 0xffff);
+            case AddressingMode.imm8ind: {
+                const index = this.bus.read16((this.state.p + 1) & 0xffff);
+                return this.bus.read(index);
+            }
+
+            case AddressingMode.imm8io: {
+                const index = this.bus.read((this.state.p + 1) & 0xffff);
+                return this.bus.read(0xff00 + index);
+            }
 
             case AddressingMode.reg8:
-            case AddressingMode.reg8_imm8:
-            case AddressingMode.reg8_ind8:
-            case AddressingMode.reg8_reg8:
-            case AddressingMode.reg8_reg8io:
-            case AddressingMode.reg8_imm8io:
-                return this.state.r8[instruction.par1];
-
-            case AddressingMode.reg16_imm16:
-            case AddressingMode.reg16:
-                return this.state.r16[instruction.par1];
+                if (par.value === undefined) {
```

Comment:
This and the following null checks are a performance overhead that will be obsolete with other changes suggested below.